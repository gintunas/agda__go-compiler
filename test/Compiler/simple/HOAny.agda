module HOAny where

open import Agda.Builtin.Nat
open import Agda.Builtin.Bool

data Wrapper (M : Set → Set) : Set where
  mnat  : M Nat → Wrapper M
  mbool : M Bool → Wrapper M

{-# FOREIGN GHC data AgdaWrapper m = Mnat (m Integer) | Mbool (m Bool) #-}
{-# COMPILE GHC Wrapper = data AgdaWrapper (Mnat | Mbool) #-}

map : ∀ {M N} → (∀ {n} → M n → N n) → Wrapper M → Wrapper N
map f (mnat mn)  = mnat (f mn)
map f (mbool mb) = mbool (f mb)

-- Higher-order use of Any in the compiled code:

-- data AgdaWrapper m = Mnat (m Integer) | Mbool (m Bool)

-- map ::
--  (() -> ()) ->
--  (() -> ()) ->
--  (() -> AgdaAny -> AgdaAny) -> AgdaWrapper AgdaAny -> AgdaWrapper AgdaAny

-- WAS:
-- The expected kind to AgdaWrapper's argument is * -> *
-- but AgdaAny's kind is *

-- WANT: SUCCESS
-- made possible by making AgdaAny poly-kinded

open import Common.Prelude

main : IO Unit
main = putStrLn ""
