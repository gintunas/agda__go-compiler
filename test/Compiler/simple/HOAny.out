EXECUTED_PROGRAM

ret > ExitSuccess
out >
out >
